proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,393 @@
+
+However, if we accept this downside as a part of a trade-off, and explore the ways that functional programming is beneficial to cryptography, there is then also the potential for functional cryptography to be made *easier*.
+
+The collective properties of referential transparency, type systems, and monadic interfaces provide a safer and more lawful context in which to write functions than the free-for-all that many imperative / OOP languages can offer, and so the functional machinery of Haskell is well-suited for adding important measures of safety and control to cryptography, helping to prevent or eliminate many issues and errors in ways that languages such as C cannot. I believe that there is room for significant improvement, given a directed effort to develop the space.
+
+# Problem Statement
+
+The Haskell cryptography ecosystem lacks significant capability beyond basic primitives. This places a significant burden on developers to properly implement various security techniques, and it exposes end users to significant risk in the event of a lapse in security. There are commercial solutions, but there is no community-driven, community-owned solution, and one is unlikely to be developed without deliberate community effort (see Appendix A - Commercial Incentive Misalignment).
+
+The Haskell cryptography ecosystem is also brittle and outdated. It is the product of considerable flux / churn over the years as various libraries have been developed and then deprecated or abandoned in favor of newer ones. The resulting libraries have highly-concrete interfaces and are strongly influenced by their implementation, both violating the abstraction barrier and preserving the design across successive libraries - as if their backend were to be changed, it would likely require changing or breaking the interface in turn.
+
+As a result, there are multiple cryptography libraries with similar interfaces, and it is not always clear which should be used preferentially. This promotes rot as downstream users unknownly rely on unmaintained libraries that slowly stop working.
+
+Furthermore, a number of important libraries in the ecosystem (eg, `cryptonite`) have recently been abandoned by their author. Their backend is written in C, which presents a long-term maintenance burden with a high knowledge and skill requirement. NaCl bindings such as `saltine` are a high-quality alternative, but solve only a limited set of problems.
```

Comment:
Are there more examples of abandoned cryptography libraries? The statement sounds unjustifiedly negative.

Reply:
I am not trying to call him out, but I am referring to Vincent Hanquez, the developer and maintainer of `cryptonite` and many other libraries, who has recently [quit Haskell entirely](https://twitter.com/vincenthz/status/1704395906374889934). The `crypton` fork had already occurred prior to this, due to a lack of maintenance and refusing a request for transfer of ownership.

Reply:
True, I just don't follow how the abandonment of Vincent's packages other than `cryptonite` is relevant to the proposal. 

There is no harm in being specific about objective facts. But vague statements of this sort can create wrong impression.

Reply:
I am unsure of how to bring it up properly without creating a strong impression. There is the original event earlier this year that lead to the forking of `crypton` and `cryptonite`, but also his recent archiving of all of his repositories, which the community has not had time to react to.

Worse, `cryptonite` [is not the only cryptography package of his](https://hackage.haskell.org/user/VincentHanquez). Most of those appear to be deprecated precursors to `cryptonite` that can be ignored, but there are several other cryptography-adjacent libraries such as `asn1-encoding` and `memory` which are not deprecated, which libraries such as `tls` and `crypton-x509` currently rely on as dependencies.

Do you have a suggestion on how to word this? This is my best attempt so far:

> Furthermore, a number of important libraries in the cryptography ecosystem (eg, `cryptonite`, `asn1-encoding`, and `memory`) have recently been archived by their author and will no longer be receiving updates.

Reply:
Sounds good to me. 

In the grand scheme of things, packages come and go all the time, it's a proper operation. If they are important enough, someone will care enough to maintain a fork.